hibernate-core/src/test/java/org/hibernate/orm/test/nationalized/CastToNationalizedStringTest.java

```diff
@@ -0,0 +1,95 @@
+package org.hibernate.orm.test.nationalized;
+
+import java.util.Locale;
+
+import org.hibernate.dialect.OracleDialect;
+
+import org.hibernate.testing.jdbc.SQLStatementInspector;
+import org.hibernate.testing.orm.junit.DomainModel;
+import org.hibernate.testing.orm.junit.JiraKey;
+import org.hibernate.testing.orm.junit.RequiresDialect;
+import org.hibernate.testing.orm.junit.RequiresDialects;
+import org.hibernate.testing.orm.junit.SessionFactory;
+import org.hibernate.testing.orm.junit.SessionFactoryScope;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.Id;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+@DomainModel(
+		annotatedClasses = {
+				CastToNationalizedStringTest.TestEntity.class
+		}
+)
+@RequiresDialects(
+		value = {
+				@RequiresDialect(value = OracleDialect.class)
+		})
+@SessionFactory(
+		statementInspectorClass = SQLStatementInspector.class
+)
+@JiraKey("HHH-16521")
+public class CastToNationalizedStringTest {
+
+	@BeforeAll
+	public void setUp(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					TestEntity entity = new TestEntity( 1l, "first" );
+					session.persist( entity );
+				}
+		);
+	}
+
+	@Test
+	public void testQuery(SessionFactoryScope scope) {
+		final SQLStatementInspector sqlStatementInspector = (SQLStatementInspector) scope.getStatementInspector();
+		sqlStatementInspector.clear();
+		scope.inTransaction(
+				session -> {
+					session.createQuery( "select CAST(t.name AS nstring(200)) from TestEntity t" )
```

Comment:
But is this legal HQL?? Where do we document that you're allowed to write this?

In HQL 6, the target type of a `cast()` is supposed to be an unqualified Java class name:

<https://docs.jboss.org/hibernate/orm/6.4/querylanguage/html_single/Hibernate_Query_Language.html#function-cast>